northstar/src/runtime/repository.rs

```diff
@@ -12,100 +12,305 @@
                 .await
-                .map_err(|e| Error::Io("Failed to remove npk".into(), e))
+                .map_err(|e| Error::io("Failed to remove npk", e))
                 .map(drop)
         } else {
-            Err(Error::InvalidContainer(container.clone()))
+            Ok(())
         }
     }
+
+    fn get(&self, container: &Container) -> Option<Arc<Npk>>
+    where
+        Self: Sized,
+    {
+        self.containers.get(container).map(|(_, npk)| npk.clone())
+    }
+
+    fn key(&self) -> Option<&PublicKey>
+    where
+        Self: Sized,
+    {
+        self.key.as_ref()
+    }
+
+    fn containers(&self) -> Vec<Arc<Npk>>
+    where
+        Self: Sized,
+    {
+        self.containers
+            .values()
+            .map(|(_, npk)| npk.clone())
+            .collect()
+    }
+}
+
+/// In memory repository
+#[derive(Default, Debug)]
+pub(super) struct MemRepository {
+    id: RepositoryId,
+    containers: HashMap<Container, (Vec<u8>, Arc<Npk>)>,
+}
+
+impl MemRepository {
+    pub(super) async fn add_buf(&mut self, buf: &[u8]) -> Result<Container, Error> {
+        let data = Vec::from(buf);
+        let fd = Self::memfd_create().map_err(|e| Error::Os("Failed create memfd".into(), e))?;
+        let file = task::block_in_place(|| {
+            let mut file = unsafe { std::fs::File::from_raw_fd(fd) };
+            file.write_all(buf)
+                .map_err(|e| Error::io("Failed copy npk", e))?;
+            file.seek(SeekFrom::Start(0))
+                .map_err(|e| Error::io("Failed seek", e))?;
+            Result::<_, Error>::Ok(BufReader::new(file))
+        })?;
+        debug!("Loading buffer");
+        let npk = npk::Npk::from_reader(file, None).map_err(|e| Error::Npk("Memory".into(), e))?;
+        let manifest = npk.manifest();
+        let container = Container::new(manifest.name.clone(), manifest.version.clone());
+
+        self.containers
+            .insert(container.clone(), (data, Arc::new(npk)));
+
+        info!("Loaded {} into {}", container, self.id);
+
+        Ok(container)
+    }
+
+    #[cfg(target_os = "android")]
+    fn memfd_create() -> nix::Result<RawFd> {
+        let name = CStr::from_bytes_with_nul(b"foo\0").unwrap();
+        let res = unsafe { nix::libc::syscall(nix::libc::SYS_memfd_create, name.as_ptr(), 0) };
+        nix::errno::Errno::result(res).map(|r| r as RawFd)
+    }
+
+    #[cfg(not(target_os = "android"))]
+    pub fn memfd_create() -> nix::Result<RawFd> {
+        let name = CStr::from_bytes_with_nul(b"foo\0").unwrap();
+        nix::sys::memfd::memfd_create(&name, nix::sys::memfd::MemFdCreateFlag::empty())
+    }
+}
+
+#[async_trait::async_trait]
+impl<'a> Repository for MemRepository {
+    async fn insert(&mut self, rx: &mut Receiver<Bytes>) -> Result<Container, Error>
+    where
+        Self: Sized,
+    {
+        let mut buffer = Vec::new();
+        while let Some(r) = rx.recv().await {
+            buffer.extend(&r);
+        }
```

Comment:
this could be moved into the function that deals with the in-memory fs (see above)

Reply:
`Repository::insert` is declared in the `Repository` trait. But the `MemRepository` needs an additional interface compared to `Repository` for loading the compiled in `hello-world` example (The npk included via `include_bytes!`. `MemRepository` needs also to satisfy `Repository` because it can also be used for installing npks via the console.